jbmc/src/java_bytecode/convert_java_nondet.cpp

```diff
@@ -29,83 +29,116 @@ Author: Reuben Thomas, reuben.thomas@diffblue.com
-  const auto next_instr=std::next(target);
-  if(!target->is_assign())
-  {
-    return next_instr;
-  }
+  const auto next_instr = std::next(target);
 
-  // Return if the rhs of the assignment isn't a side effect expression
-  const auto &assign=to_code_assign(target->code);
-  if(assign.rhs().id()!=ID_side_effect)
+  // We only expect to find nondets in the rhs of an assignments or in return
+  // statements
+  for(exprt &op : target->code.operands())
   {
-    return next_instr;
-  }
+    if(!can_cast_expr<side_effect_expr_nondett>(op))
+    {
+      continue;
+    }
 
-  // Return if the rhs isn't nondet
-  const auto &side_effect=to_side_effect_expr(assign.rhs());
-  if(side_effect.get_statement()!=ID_nondet)
-  {
-    return next_instr;
-  }
+    const auto &nondet_expr = to_side_effect_expr_nondet(op);
 
-  const auto lhs=assign.lhs();
-  // If the lhs type doesn't have a subtype then I guess it's primitive and
-  // we want to bail out now
-  if(!lhs.type().has_subtype())
-  {
-    return next_instr;
-  }
+    // If the lhs type doesn't have a subtype then I guess it's primitive and
+    // we want to bail out now. If the type is a ptr-to-void or a function
+    // pointer then we also want to bail.
+    const typet &type = nondet_expr.type();
+    if(
+      type.id() != ID_pointer || type.subtype().id() == ID_empty ||
+      type.subtype().id() == ID_code)
+    {
+      continue;
+    }
 
-  // Although, if the type is a ptr-to-void then we also want to bail
-  if(lhs.type().subtype().id()==ID_empty ||
-     lhs.type().subtype().id()==ID_code)
-  {
-    return next_instr;
+    if(!nondet_expr.get_nullable())
+      object_factory_parameters.max_nonnull_tree_depth = 1;
+
+    const auto source_loc = target->source_location;
+
+    // Currently the code looks like this
+    //   target: original instruction
+    // When we are done it will look like this
+    //   target : declare aux_symbol
+    //      .     <gen_nondet_init_instructions - many lines>
+    //      .     <gen_nondet_init_instructions - many lines>
+    //      .     <gen_nondet_init_instructions - many lines>
+    //   target2: orig instruction with op replaced by aux_symbol
+    //            dead aux_symbol
+
+    symbolt &aux_symbol = get_fresh_aux_symbol(
+      type,
+      id2string(goto_programt::get_function_id(goto_program)),
+      "nondet_tmp",
+      source_loc,
+      ID_java,
+      symbol_table);
+    aux_symbol.is_static_lifetime = false;
+
+    const symbol_exprt aux_symbol_expr = aux_symbol.symbol_expr();
+    op = aux_symbol_expr;
+
+    // It is simplest to insert the final instruction first, before everything
+    // gets moved around
+    goto_programt::targett dead_aux_symbol = goto_program.insert_after(target);
+    dead_aux_symbol->type = DEAD;
+    dead_aux_symbol->code = code_deadt(aux_symbol_expr);
+    dead_aux_symbol->source_location = source_loc;
+
+    // Insert an instruction declaring `aux_symbol` before `target`, being
```

Comment:
Nit: add the decl before adding the dead, (don't think it makes any different since target remains unmoved after each op?)

Reply:
Nope. After `goto_program.insert_before_swap(target, decl_aux_symbol);`, `target` points to the `decl_aux_symbol` and `std::next(target)` points to what `target` used to point to. This is because what we actually do is put `decl_aux_symbol` in a new instruction after `target` and then swap the contents of `target` and `std::next(target)`.

However, I think I can make it a bit simpler.